agent/xdsv2/rbac_resources.go

```diff
@@ -0,0 +1,246 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package xdsv2
+
+import (
+	"fmt"
+
+	envoy_listener_v3 "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	envoy_rbac_v3 "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
+	envoy_route_v3 "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	envoy_network_rbac_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/rbac/v3"
+	envoy_matcher_v3 "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+
+	"github.com/hashicorp/consul/agent/xds/response"
+	"github.com/hashicorp/consul/proto-public/pbmesh/v1alpha1/pbproxystate"
+)
+
+const (
+	baseL4PermissionKey = "consul-intentions-layer4"
+)
+
+func MakeL4RBAC(trafficPermissions *pbproxystate.L4TrafficPermissions) ([]*envoy_listener_v3.Filter, error) {
+	if trafficPermissions == nil {
+		return nil, nil
+	}
+
+	rbacs, err := makeRBACs(trafficPermissions)
+	if err != nil {
+		return nil, err
+	}
+
+	filters := make([]*envoy_listener_v3.Filter, 0, len(rbacs))
+	for _, rbac := range rbacs {
+		cfg := &envoy_network_rbac_v3.RBAC{
+			StatPrefix: "connect_authz",
+			Rules:      rbac,
+		}
+		filter, err := makeEnvoyFilter("envoy.filters.network.rbac", cfg)
+		if err != nil {
+			return nil, err
+		}
+		filters = append(filters, filter)
+	}
+
+	return filters, nil
+}
+
+func makeRBACs(trafficPermissions *pbproxystate.L4TrafficPermissions) ([]*envoy_rbac_v3.RBAC, error) {
+	allowRBAC := &envoy_rbac_v3.RBAC{
+		Action:   envoy_rbac_v3.RBAC_ALLOW,
+		Policies: make(map[string]*envoy_rbac_v3.Policy),
+	}
+
+	denyRBAC := &envoy_rbac_v3.RBAC{
+		Action:   envoy_rbac_v3.RBAC_DENY,
+		Policies: make(map[string]*envoy_rbac_v3.Policy),
+	}
+
+	policyLabel := func(i int) string {
+		if len(trafficPermissions.Permissions) == 1 {
+			return baseL4PermissionKey
+		}
+		return fmt.Sprintf("%s-%d", baseL4PermissionKey, i)
+	}
+
+	for i, p := range trafficPermissions.Permissions {
+		allowPolicy, err := makeRBACPolicy(allowRBAC.Action, p.AllowPrincipals)
+		if err != nil {
+			return nil, err
+		}
+
+		if allowPolicy != nil {
+			allowRBAC.Policies[policyLabel(i)] = allowPolicy
```

Comment:
I was imagining that we would instead put each `AllowPrincipal` into a policy with the index of that `AllowPrincipal` object. This would allow us to avoid constructing the `orPrincipal` for all allow/deny principals.